Add codespell support (config, workflow to detect/not fix) and make it fix few typos

yarikoptic opened the pull request:
More about codespell: https://github.com/codespell-project/codespell .

I personally introduced it to dozens if not hundreds of projects already and so far only positive feedback.

CI workflow has 'permissions' set only to 'read' so also should be safe.

- There were numerous commits where `codespell` was used "manually" to fixup some of the typos.
- https://github.com/numpy/numpy/issues/13694 still contemplates idea of running `codespell` on PRs
- An early attempt to introduce codespell CI was abandoned: https://github.com/numpy/numpy/pull/19594 -- IMHO now is better time!

- I did `lapack_lite` last in the separate commits if there is desire to not touch it (although I saw commits fixing typos in it as well).
- I did some variable renames to take them away from "close to a typo" form, might be worth having a look and suggesting alternative names.
- I did fix older changelogs as I think there is no point to cherish typos in them, but they could also be ignored.

yarikoptic committed:
Add rudimentary codespell config

yarikoptic committed:
For now exclude lapack_lit from spelling

yarikoptic committed:
Add some common hits to be ignored by codespell

Used
  codespellit list-hits-sorted | sed -n -e '/6 /,9999p' | tac | f2 | tr '\n' ','
to find and sort

yarikoptic committed:
Ignore names and acronyms

yarikoptic committed:
Rename "datas" to "data_samples"

yarikoptic committed:
more skips

yarikoptic committed:
tune some variables or annotate to be skipped to avoid triggering codespell + parms -> params

yarikoptic committed:
Annotate line with secnd - I think on purpose

yarikoptic committed:
Replace typo-close doub with 'd' in an example

yarikoptic committed:
DOC: adjust variables in example to be fp_{kind} instead of {k}fp and fp{k}

I think it makes them more consistent (always a suffix) and also more
readable since more descriptive of the taken action.

This is triggered to avoid "looks-like-a-typo" hits  in codespell for fpr and likely
others

yarikoptic committed:
Unclear typo fix guess

yarikoptic committed:
Replace typo-close padd with arr_ (name padd seems to not carry semantic load there)

yarikoptic committed:
Mark intentional typo to be skipped

yarikoptic committed:
[DATALAD RUNCMD] Do interactive fixing of  leftover ambigous typos

=== Do not change lines below ===
{
 "chain": [],
 "cmd": "codespell -w -i 3 -C 2",
 "exit": 0,
 "extra_inputs": [],
 "inputs": [],
 "outputs": [],
 "pwd": "."
}
^^^ Do not change lines above ^^^

yarikoptic committed:
Re-enable codespelling lapack_lite and skip few common short variable names

yarikoptic committed:
[DATALAD RUNCMD] run codespell throughout for lapack_lite fixing typo automagically but ignoring the failure due to ambigous typos

=== Do not change lines below ===
{
 "chain": [],
 "cmd": "codespell -w || :",
 "exit": 0,
 "extra_inputs": [],
 "inputs": [],
 "outputs": [],
 "pwd": "."
}
^^^ Do not change lines above ^^^

yarikoptic committed:
[DATALAD RUNCMD] Do interactive fixing of lapack_lite leftover ambigous typos

=== Do not change lines below ===
{
 "chain": [],
 "cmd": "codespell -w -i 3 -C 2",
 "exit": 0,
 "extra_inputs": [],
 "inputs": [],
 "outputs": [],
 "pwd": "."
}
^^^ Do not change lines above ^^^

charris commented:
Please don't touch the changelogs, they match the title used in the corresponding PR, typos and all.

ngoldbaum commented on doc/source/dev/governance/governance.rst:
```diff
@@ -262,7 +262,7 @@ conflict of interests include, but are not limited to:
 conflict of interest they may have. Members with a conflict of interest
 in a particular issue may participate in Council discussions on that
-issue, but must recuse themselves from voting on the issue.
+issue, but must recurse themselves from voting on the issue.
```
not a big fan of a adding automated spell-checking if we'll have to ignore it making mistakes like this...

yarikoptic replied:
rright -- I even ran into this particular one before -- `recuse` is indeed rare and I failed to spot it in review, sorry about that. I will whitelist it.

yarikoptic committed:
ignore changelogs and "recuse"

yarikoptic committed:
replace typo-close "notin" with "not_in" variable

yarikoptic committed:
fix ambiguous "fals -> false" typo

yarikoptic committed:
Whitelist implementor for codespell

yarikoptic committed:
Fixing linter's gotcha on spaces before operator

Funny thing is that it does not catch it on 'main' for some reason
but catches here... it is likely because

  python tools/linter.py --branch origin/main

works on the diff only (too fast on main) and thus whenever those
spaces were introduced, it did not check and never flipped since then.
I changed that line, and it was checked and flagged.

yarikoptic committed:
Use _ for unused return value (makes line shorter to pass lint)

yarikoptic commented:
Thanks for quick feedback @charris and @ngoldbaum !

I
- rebased (apparently there were new typos in main)
- ran with current `master` of codespell to ensure picking up as many typos as possible (thought that typos detected by CI were due to mine being outdated but it just needed rebase)
- whitelisted some more per feedback above (do not touch changelogs, `recuse` is ok so we might `recuse into folders` at some point as a side-effect ;-) )
- adjusted some patches to make `lint` satisfied (one change with spaces around `=` was for "old" code, but triggered to be checked since I changed the line)

yarikoptic commented on doc/changelog/1.25.0-changelog.rst:
```diff
@@ -563,7 +563,7 @@ A total of 530 pull requests were merged for this release.
 * `#23566 <https://github.com/numpy/numpy/pull/23566>`__: CI: disable intel_spr_sde_test for now
 * `#23568 <https://github.com/numpy/numpy/pull/23568>`__: MAINT: remove usages of sys.exc_info
-* `#23569 <https://github.com/numpy/numpy/pull/23569>`__: TST: try readding test_new_policy on musl
+* `#23569 <https://github.com/numpy/numpy/pull/23569>`__: TST: try re-adding test_new_policy on musl
```
oh -- missed this one, was probably modified by another than "fix all" commit... will fix up now...

mattip commented on numpy/_core/memmap.py:
```diff
@@ -162,48 +162,48 @@ class memmap(ndarray):
-    >>> newfp = np.memmap(filename, dtype='float32', mode='r', shape=(3,4))
-    >>> newfp
+    >>> fp_new = np.memmap(filename, dtype='float32', mode='r', shape=(3,4))
+    >>> fp_new
```
Did codespell do this? It seems unnecessary

yarikoptic replied:
nope, I did that in 639693ee4fa82173f5444e76cfec47704d655ea6 :

```
commit 639693ee4fa82173f5444e76cfec47704d655ea6
Author: Yaroslav Halchenko <debian@onerussian.com>
Date:   Fri Apr 19 10:02:05 2024 -0400

    DOC: adjust variables in example to be fp_{kind} instead of {k}fp and fp{k}
    
    I think it makes them more consistent (always a suffix) and also more
    readable since more descriptive of the taken action.
    
    This is triggered to avoid "looks-like-a-typo" hits  in codespell for fpr and likely
    others
```

mattip commented on numpy/_core/memmap.py:
```diff
@@ -162,48 +162,48 @@ class memmap(ndarray):
-    >>> fpr = np.memmap(filename, dtype='float32', mode='r', shape=(3,4))
-    >>> fpr.flags.writeable
+    >>> fp_ro = np.memmap(filename, dtype='float32', mode='r', shape=(3,4))
+    >>> fp_ro.flags.writeable
```
?

yarikoptic replied:
see above... too close to `for` typo etc.

mattip commented on numpy/_core/tests/test_einsum.py:
```diff
@@ -816,7 +816,7 @@ def test_einsum_fixedstridebug(self):
         assert_equal(es,  tp)
         # The following is the original test case from the bug report,
-        # made repeatable by changing random arrays to aranges.
+        # made repeatable by changing random arrays to arranges.
```
revert this

yarikoptic replied:
good catch, adding `arange` and `aranges` to whitelist

mattip commented on numpy/_core/tests/test_indexing.py:
```diff
@@ -534,8 +534,8 @@ def test_indexing_array_weird_strides(self):
-        arro = np.zeros((4, 4))
-        arr = arro[::-1, ::-1]
+        arr0 = np.zeros((4, 4))
+        arr = arr0[::-1, ::-1]
```
?

yarikoptic replied:
see above and specifically commit  8afd1bf9e4aac96eb3e91f62c0b5ed7bd0b2f495

mattip commented on numpy/_core/tests/test_multiarray.py:
```diff
@@ -4810,8 +4810,8 @@ def test_combinations(self, data):
-        padd = np.repeat(np.min(arr), 513)
-        rarr = np.concatenate((arr, padd))
+        arr_ = np.repeat(np.min(arr), 513)
+        rarr = np.concatenate((arr, arr_))
```
?

yarikoptic replied:
7c5605731885495dc06f035439cfa5bfaa1bf85e

mattip commented on numpy/_core/tests/test_simd.py:
```diff
@@ -900,8 +900,8 @@ def test_misc(self):
-            broadcasti = self.setall(i)
-            assert broadcasti == [i] * self.nlanes
+            broadcast_i = self.setall(i)
+            assert broadcast_i == [i] * self.nlanes
```
?

yarikoptic replied:
8afd1bf9e4aac96eb3e91f62c0b5ed7bd0b2f495

mattip commented on numpy/_pyinstaller/hook-numpy.py:
```diff
@@ -18,7 +18,7 @@
     # collected to capture MKL, OpenBlas, OpenMP, etc.
     from PyInstaller.utils.hooks import conda_support
-    datas = conda_support.collect_dynamic_libs("numpy", dependencies=True)
+    _ = conda_support.collect_dynamic_libs("numpy", dependencies=True)
```
No need to have a left-side at all

mattip commented on numpy/lib/_shape_base_impl.py:
```diff
@@ -37,7 +37,7 @@ def _make_along_axis_idx(arr_shape, indices, axis):
     dest_dims = list(range(axis)) + [None] + list(range(axis+1, indices.ndim))
 
-    # build a fancy index, consisting of orthogonal aranges, with the
+    # build a fancy index, consisting of orthogonal arranges, with the
```
Revert

mattip commented on pyproject.toml:
```diff
@@ -213,3 +213,14 @@ cli = 'vendored-meson/meson/meson.py'
+# Skip all shortish (up to 5) ACRONYMS
+ignore-regex = '"worl"|https://citeseerx\.ist\.psu\.edu|\b([A-Z][a-z]{1,9}|[A-Z]{1,5})\b|.*codespell-ignore.*|/seh/'
+# some common hits and some "questionable" and false positives (rare words)
+ignore-words-list = 'nd,nin,numers,numer,inout,ccompiler,fo,CCompiler,dout,ans,ser,ot,asign,siz,alow,fro,delimitor,alo,dum,desig,recuse,implementor'
```
Can this be folded into a multiline statement?

yarikoptic replied:
Ha -- it seems to can. Splitting into multiple lines

mattip commented on doc/source/release/1.14.0-notes.rst:
```diff
@@ -334,7 +334,7 @@ eliminating their use internally and two new C-API functions,
 have been added together with a complementary flag,
 ``NPY_ARRAY_WRITEBACKIFCOPY``. Using the new functionality also requires that
-some flags be changed when new arrays are created, to wit:
+some flags be changed when new arrays are created:
```
Please leave these files out.

yarikoptic replied:
skipping and dropping manual 21d95b25f5ae2c09e5cc6f55bbcf785aa957c8dd

mattip reviewed (commented):
On the one hand, this found many typos. On the other, it has definite opinions about variable names in tests, not all are valid. This is kind of like black in that it operates on the entire code base, and like black seems to be over-strict.

yarikoptic committed:
Add arange and aranges to whitelist

yarikoptic committed:
Place skips for codespell into multiple lines for easier digestion

yarikoptic committed:
Do not codespell ./doc/source/release/*-notes.rst

yarikoptic commented:
commented on above review by @mattip , dropped a commit, rebased the latter ones, reran codespell (so now no changes to release notes), force pushing now.

yarikoptic committed:
do not codespell recarray_from_file.fits - just raises warnings

yarikoptic committed:
[DATALAD RUNCMD] run codespell throughout fixing typos automagically

=== Do not change lines below ===
{
 "chain": [],
 "cmd": "codespell -w",
 "exit": 0,
 "extra_inputs": [],
 "inputs": [],
 "outputs": [],
 "pwd": "."
}
^^^ Do not change lines above ^^^

mattip commented:
Can you split this into 2 PRs: one to fix the spelling errors and another to add a codespell CI job? Then we can merge the spelling fixes while still debating the need for another CI job vs. a pre-commit hook.

yarikoptic commented:
Sure!  If you don't mind, I will also keep configuration for codespell along with the PR fixing typos.

Note: I do not even see you using pre-commit here -- would be happy to add hook there too/instead

yarikoptic commented:
I force pushed this branch kicking action away.  After it is merged, will propose PR with the action. If to propose now - I would need to base it on this branch or it would lead to failed CI run.
